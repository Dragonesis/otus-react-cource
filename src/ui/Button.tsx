import styled from '@emotion/styled'
import { css } from '@emotion/react'

export interface ButtonProps {
  size?: 's' | 'm' | 'l'
  fixSize?: 's' | 'm' | 'l'
  variant?: 'dark' | 'light'
}

const dark = () =>
  css`
    background-color: #3a3a3a;
    color: #fff;
  `

const light = () =>
  css`
    background-color: #fff;
    color: #3a3a3a;
  `
const sizeS = () =>
  css`
    height: 36px;
    padding-left: 10px;
    padding-right: 10px;
    font-size: 16px;
  `

const sizeM = () =>
  css`
    height: 48px;
    padding-left: 10px;
    padding-right: 10px;
    font-size: 18px;
  `

const sizeL = () =>
  css`
    height: 60px;
    padding-left: 16px;
    padding-right: 16px;
    font-size: 20px;
  `
const sizeMap = {
  s: sizeS,
  m: sizeM,
  l: sizeL,
}
const fixSizeMap = {
  s: '180px',
  m: '210px',
  l: '250px',
}
const variantMap = {
  dark,
  light,
}

export const Button = styled.button<ButtonProps>`
  border-radius: 6px;
  width: 100%;
  max-width: ${(props) => props.fixSize && fixSizeMap[props.fixSize]};

  &:hover,
  &:focus {
    border-color: #303030;
  }

  ${(props) => {
    return [props.size ? sizeMap[props.size] : sizeS, props.variant ? variantMap[props.variant] : dark]
  }}

  &.size-s {
    height: 36px;
    padding-left: 10px;
    padding-right: 10px;
    font-size: 16px;
  }

  &.size-m {
    height: 48px;
    padding-left: 10px;
    padding-right: 10px;
    font-size: 18px;
  }

  &.size-l {
    height: 60px;
    padding-left: 16px;
    padding-right: 16px;
    font-size: 20px;
  }
`
