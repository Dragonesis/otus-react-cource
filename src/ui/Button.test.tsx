import React from 'react'
import { render } from '@testing-library/react'

import { Button } from './Button'

describe('### Button', () => {
  test('Button renders correctly', () => {
    const { asFragment } = render(<Button />)
    expect(asFragment()).toMatchSnapshot()
  })
})
