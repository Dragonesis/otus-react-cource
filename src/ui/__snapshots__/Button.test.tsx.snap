// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`### Button Button renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  border-radius: 6px;
  width: 100%;
  height: 36px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 16px;
  background-color: #3a3a3a;
  color: #fff;
}

.emotion-0:hover,
.emotion-0:focus {
  border-color: #303030;
}

.emotion-0.size-s {
  height: 36px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 16px;
}

.emotion-0.size-m {
  height: 48px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 18px;
}

.emotion-0.size-l {
  height: 60px;
  padding-left: 16px;
  padding-right: 16px;
  font-size: 20px;
}

<button
    class="emotion-0"
  />
</DocumentFragment>
`;
